keps/sig-auth/2579-pod-isolation-policy/README.md

```diff
@@ -0,0 +1,951 @@
+should policy versions past the cluster version be handled? Should `latest` mean the latest version
+supported by the webhook, or the policy version matching the cluster version? _Leaning towards
+supporting all policy versions supported by the webhook, even if they are newer than the cluster
+version._
+
+Note that policies are not guaranteed to be backwards compatible, and a newer restricted policy
+could require setting a field that doesn't exist in the current API version.
+
+<<[/UNRESOLVED]>>
+
+<<[UNRESOLVED]>>
+
+- Under an older version of a policy, a new field may only be set to its default value (or left unset).
+
+**Open Question:** Should this apply to all new fields, or only new fields that are constrained
+under a more recent version of the policy? If we only constrain fields that are in-scope for policy
+enforcement, then that becomes a problem if we later decide that a field should be in scope. The
+versioning machanism gives us a way to safely make "breaking" changes like adding enforcement to a
+previously unenforced field, but it is strictly less safe. Either way, there is some maintenance
+burden to declare which fields are in-scope at each version. _Leaning towards only restricting
+relevant fields._
+
+**Open Question:** If the control (or options) are strictly less privileged than the default, should
+it be allowed? For example, the `allowPrivilegeEscalation` field was added in Kubernetes v1.8, which
+sets the `no_new_privs` flag on the container processes when set to false. Prior to that version,
+the `no_new_privs` flag was never set, which is also the default in v1.8+. If older policies require
+the default value, then only `allowPrivilegeEscalation: true` would be allowed, even though
+`allowPrivilegeEscalation: false` is strictly less privileged. _Leaning towards allowing strictly
+more-restricted versions._
+
+- If the range of valid values on an existing field is expanded, the new values are forbidden. E.g.
+  adding a new enum value.
+
+Same questions as above
+
+<<[/UNRESOLVED]>>
+
+### PodTemplate Resources
+
+Audit and Warn modes are also checked on resource types that embed a PodTemplate (enumerated below),
+but Allow mode only applies to actual pod resources.
+
+Since users do not create pods directly in the typical deployment model, the warning mechanism is
+only effective if it can also warn on templated pod resources. Similarly, for audit it is useful to
+tie the audited violation back to the requesting user, so audit will also apply to templated pod
+resources. In the interest of supporting mutating admission controllers, policies will only
+be enforced on actual pods.
+
+Templated pod resources include:
+
+- v1 ReplicationController
+- v1 PodTemplate
+- apps/v1 ReplicaSet
+- apps/v1 Deployment
+- apps/v1 StatefulSet
+- apps/v1 DaemonSet
+- batch/v1 CronJob
+- batch/v1 Job
+
+CRDs that wish to take advantage of this functionality should instead have an object reference to a
```

Comment:
This seems too high of a bar.  How many workload extensions actually support pod template today?  I’d expect us to have a better story here before GA, but this seems a high bar on the surface.  Pod template isn’t our “default workload pattern” in or out of tree.  It seems much more reasonable to have a better way to register resources for consideration by the central admission plug-in.

Reply:
Can we have a way to "mark" a given workload type is `podspecable`?

Reply:
I *really* don't want to special-case pod specs in CRDs. In general, I think referencing PodTemplate is superior to inlining a point-in-time version of the podspec API into a CRD schema.

Reply:
Also - this is sort of "nice to have" friendly UX, the actual enforcement still occurs on the pod. If you want the audit & warn capabilities on the CRD, then you can always ship a custom webhook with the CRD.

Reply:
> . In general, I think referencing a PodTemplate is superior to inlining a point-in-time version of the podspec API into a CRD schema.

i don’t think you have the data to back up that’s what users want (no one has ever told me they want to update a pod template vs deployment).

i agree it may be theoretically better but i absolutely expect the way we support pod templates is inline by default because that’s how kube works, preserves atomicity, and as far as i know no one actually uses pod templates like this (but happy to be proven wrong)

Reply:
A lot of the ways kube works by default with inline pod specs does *not* work with inline pod specs in CRs:
* 100% fidelity with pod spec fields known to the API server
* matching validation for the template pod spec and the resulting pod (unique container names, resource validation, etc)
* support for client-side apply

Regardless, since this isn't core to the enforcement mechanism, it's reasonable to scope the in-tree mechanism specific to pod specs to operate on in-tree types containing pod specs, especially since one of those types is an inert type that *can* be consumed by out of tree extensions if desired.

Reply:
Rephrased. With the options described here.

I'm open to considering an inlined CRD implementation in a future enhancement of this feature, but not in the initial version.